File: overlay/tests/test_bogus.py
"""
test_bogus.py - Test server error handling for malformed/bogus commands.
"""

import os
import sys
import unittest

sys.path.insert(0, os.path.dirname(__file__))
from memcached_harness import MemcachedServer


class TestBogusCommands(unittest.TestCase):
    @classmethod
    def setUpClass(cls):
        cls.server = MemcachedServer().start()

    @classmethod
    def tearDownClass(cls):
        cls.server.stop()

    def setUp(self):
        self.mc = self.server.client()

    def tearDown(self):
        self.mc.close()

    def test_unknown_command(self):
        self.mc.send_cmd("boguscommand")
        res = self.mc.readline().strip()
        self.assertEqual(res, "ERROR")

    def test_malformed_set(self):
        # Missing arguments in set
        self.mc.send_cmd("set foo 0")
        res = self.mc.readline().strip()
        self.assertTrue(res.startswith("CLIENT_ERROR") or res.startswith("ERROR"))

    def test_non_numeric_flags(self):
        self.mc.send_cmd("set foo notanumber 0 5\r\nhello")
        res = self.mc.readline().strip()
        self.assertTrue("ERROR" in res)

    def test_negative_bytes(self):
        self.mc.send_cmd("set foo 0 0 -1\r\nhello")
        res = self.mc.readline().strip()
        self.assertTrue("ERROR" in res)


if __name__ == "__main__":
    unittest.main()

File: overlay/tests/test_bogus.c
/*
 * test_bogus.c - Test server error handling for malformed/bogus commands.
 */

#include "c_harness.h"

static int test_unknown_command(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char line[128];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_send_str(&c, "boguscommand\r\n") == 0);
  TEST_ASSERT(client_readline(&c, line, sizeof(line)) > 0);
  TEST_ASSERT(strstr(line, "ERROR") != NULL);

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

static int test_malformed_set(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char line[128];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_send_str(&c, "set foo 0\r\n") == 0);
  TEST_ASSERT(client_readline(&c, line, sizeof(line)) > 0);
  TEST_ASSERT(strstr(line, "CLIENT_ERROR") != NULL ||
              strstr(line, "ERROR") != NULL);

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

static int test_non_numeric_flags(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char line[128];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_send_str(&c, "set foo notanumber 0 5\r\nhello\r\n") == 0);
  TEST_ASSERT(client_readline(&c, line, sizeof(line)) > 0);
  TEST_ASSERT(strstr(line, "ERROR") != NULL);

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

static int test_negative_bytes(void) {
  struct memcached_srv srv;
  struct memcached_client c;
  char line[128];

  TEST_ASSERT(srv_start(&srv, 0, NULL) == 0);
  TEST_ASSERT(client_connect(&c, srv.port) == 0);

  TEST_ASSERT(client_send_str(&c, "set foo 0 0 -1\r\nhello\r\n") == 0);
  TEST_ASSERT(client_readline(&c, line, sizeof(line)) > 0);
  TEST_ASSERT(strstr(line, "ERROR") != NULL);

  client_close(&c);
  srv_stop(&srv);
  return 0;
}

int test_bogus_run(void) {
  RUN_TEST(test_unknown_command);
  RUN_TEST(test_malformed_set);
  RUN_TEST(test_non_numeric_flags);
  RUN_TEST(test_negative_bytes);
  return 0;
}
